### DISCUSSION TITLE
Support client: directives on <script> tags

### DISCUSSION THREAD

**Discussion by zadeviggers** [withastro/roadmap#102] [Proposal]
Sometimes I want to add a teeny-tiny bit of interactivity to my site without having to add all the extra weight that comes with a whole front-end framework. This can easily be done using `<script>` tags inside of `.astro` files, but has the downside of not being able to use all of the `client:` directives available when using components from frameworks.
What I propose is allowing all of the existing `client:` directives (except maybe `client:only` - not sure how that would work) to be used on `<script>` tags inside of `.astro` files.
The current approach of only supporting the `client:` directives on components for fully-fledged front-end framework feels like it treats vanilla JavaScript as a second-class citizen, which just feels wrong.

I also considered supporting the `client:` directives on Astro components instead of just `<script>` tags, but decided against it because of the confusion that could create around what does/doesn't run on the server, and it would also make using the directives on scripts at the top level inside of routing `.astro` files (the ones in `src/pages/`) impossible.

**Comment by matthewp** [comment]
I like this idea!

**Reply by mariohamann** [reply]
Hey there,

I believe `client-*`-directives are really a great way to improve performance and sometimes even to handle functionality (e g. load and execute all tracking scripts when `client:tracking` ran...)

To explore ways I've created a package, that allows to use more or less usual script tags with `client:*` directives. This is how it currently works:


```astro
---
import ScriptIsland from 'astro-script-islands/component';
---

<ScriptIsland client:idle>
  <script>
    console.log('This script loads when the browser is idle!');
  </script>
</ScriptIsland>
```

Check out the [docs and demos](https://mariohamann.github.io/astro-script-islands/) or the [code](https://github.com/mariohamann/astro-script-islands) for some insights.

My main idea was: I'm actually faking a framework and therefore have some control over client and server side rendering. All `client:*`- directives work, even custom ones. There are some problems, e. g. that you can't currently have scripts with and without `<ScriptIsland>` in one component file. In general the implementation process was horrible and I would MORE than highly appreciate if this could become first party functionality But for the moment I'm happy with the result.

Feel free to reach out, if you have questions. :)

**Comment by aFuzzyBear** [comment]
Im certain this was asked a while ago, mind you there is xelement that would help take care of this sort of issue, where you can apply your code to `@visible` `@resize` or even have it use the DOM mutation observer `@observe` to fire on certain dom changes, 
I alway felt that `<script>` tags are sacrosanct in the respective fact that they are the gateway into JS-land on the client, 
Partial Hydration is a technique used to help with the loading and execution of JS components, built using UI frameworks, where the `astro-root` gets wrapped around your island. This proposal would then suggest one further that the script is no longer sacrosanct instead each `<script>` is now wrapped under an `astro-root`, this doesnt feel like good html practice to me, personally I would like to see Astro go the other way and have something similar to `qwik` with their streaming of JS using html attributes for Server-Side Required Content, but that is a seperate PR,

**Comment by delucis** [comment]
I think I’d never seen this discussion although this is something I’ve also wanted before and chatted about with people!

I think one challenge will be that some of our hydration hooks like `client:visible` rely on a component’s physical position on a page, which a script tag doesn’t really have. One option here would be instead to support the `client:*` directive on the Astro component usage itself rather than directly on `<script>` (i.e. `<AstroComponent client:visible />` would only load the contained scripts when visible).

Another option I’ve discussed with people before is exposing loading utilities for people to do this manually which is simpler to implement on the Astro side although it’s a less nice experience for users as you probably need to move your script to an external file to benefit:

```html
<script>
import { onLoad, onMedia, onIdle } 'astro/client-directives';

onLoad(() => import('./bigger-script'));
onIdle(() => import('./bigger-script'), { timeout: 2000 });
onMedia(() => import('./bigger-script'), '(max-width: 50em)');
</script>
```

`client:visible` would probably need users to manually specify a selector in this case, like this for example:

```html
<custom-element></custom-element>

<script>
import { onVisible } 'astro/client-directives';

onVisible(() => import('./bigger-script'), { target: 'custom-element', rootMargin: 200px });
</script>
```

Since Astro 2.6.0 there is also support for [custom directives](https://docs.astro.build/en/reference/directives-reference/#custom-client-directives), so it could also be that the import should be from a virtual module to support those:

```js
import { onRainyInNewYork } from 'astro:client-directives';
```

That said, as @zadeviggers also [noted](https://github.com/withastro/roadmap/discussions/102#discussioncomment-12608429) below, the DX improvement here is only small. `onIdle()` for example would be pretty much the same as `requestIdleCallback()`, and other directives only require moderately more code.

**Reply by jasikpark** [reply]
It would be pretty cool to implement the loaders for web components. Like somehow pipe `client:load` into the web component callback when using Astro components wrapping a simple web component? I'm unsure of if there's an actual case where that helps.

**Reply by delucis** [reply]
We do something a bit like this for Starlight’s search component: https://github.com/withastro/starlight/blob/a94cccc3185e4015e0f5acd15ce542d8750d4c92/packages/starlight/components/Search.astro#L139-L141

The search input is inside a modal and we only load the actual search library on idle. I guess a change like this could allow us to refactor it to do something like `<SearchForm client:idle />` rather than writing our own idle callback.

**Reply by zadeviggers** [reply]
Wow, I asked about this 3 years ago. Time flies!

Looking at it, I agree that it makes the most sense to have this capability for Astro components (+ maybe web components) and the scripts within them rather than script tags themselves. Exposing the directives in a virtual module could be useful in some situations, but at that point, the difference from doing it yourself with an IntersectionObserver/whatever is quite minimal.
